Move `from_email_address` options to new SiteEmailAddresses entity

colemanw opened the pull request:
Overview
----------------------------------------
Moves the 'from_email_address' option group into its own table: Introducing the new `SiteEmailAddress` entity. 🎉

Migration and Backward Compatibility Notes
-----------

### New-to-Old column map

| SiteEmailAddress | OptionValue |
-- | --
`id` | `value`
`display_name` | `label` (in quotes)
`email` | `label` (in brackets)
`description` | `description`
`is_active` | `is_active`
`is_default` | `is_default`
`domain_id` | `domain_id`

### Backward Compatibility

_This adds an apiWrapper for backward-compatibility, to allow `SiteEmailAddress` records to be accessed as if it were an option group. When accessing it via the OptionValue api, the new `email` and `display_name` columns are concatenated to produce familiar `value` of `"Name" <email@.example.com>`, and `id` is returned as `value`. Here's what is and is not supported:_

| Function | Works | Notes |
 -- | -- | --
**Api3 get** | ✅ | `$params` must contain `'option_group_id' => 'from_email_address'`
**Api3 create** | ✅ | `$params` must contain `'option_group_id' => 'from_email_address'`
**Api3 update** (aka. `create` + an `id`) | ⚠️ |  works if passed `id` of an existing siteEmailAddress and a correctly formatted email header as `value`
**Api3 delete** | 🔴 | `delete` action only gets passed an `id` which is not enough info for the adapter
**Api3 chaining** | ✅ | `get` chained with create/update/delete will all work
**Api4 get** | ✅ | `WHERE` must contain `'option_group_id:name', '=', 'from_email_address'`
**Api4 update** | ⚠️ | Update by id won't work. `WHERE` must contain `'option_group_id:name', '=', 'from_email_address'`
**Api4 create** | ✅ | `$values` must contain `'option_group_id:name' => 'from_email_address'`
**Api4 delete** | ⚠️ | Delete by id won't work. `WHERE` must contain `'option_group_id:name', '=', 'from_email_address'`
**Api4 chaining** | 🔴 | Couldn't find any uses in `universe` so didn't add support for this
**`CRM_Core_BAO_Domain::getNameAndEmail`**  **`CRM_Core_BAO_Domain::getFromEmail`** | ✅ | No signature change
**`CRM_Core_BAO_Email::getFromEmail()`** **`CRM_Core_BAO_Email::domainEmails()`**  | ✅ | No signature change
**`CRM_Core_OptionGroup::values`** | ✅ | Supported params: `$labelColumnName`, `$keyColumnName`, `$onlyActive`, `$condition`
**`CRM_Core_OptionGroup::getDefaultValue`** | ✅ | Returns `id` of default email address for current domain
**Other core functions** | 🔴 | `CRM_Core_OptionGroup::valuesByID`, `CRM_Core_Pseudoconstant::get`, etc.
**SQL Queries, etc.** | 🔴 | The options no longer exist in the `civicrm_option_value` table so can't be queried.

Before
--------
The list of from email addresses was not well-placed in the OptionValue table because:
1. It was the only domain-specific option, overcomplicating that table with domain handling just for this one option list (aside, there was actually one other: _Grant Type_, but [probably wasn't actually being used for multidomain](https://github.com/civicrm/civicrm-core/pull/31924)).
2. Had to mush the display name and email address together, like `"My Name" <myname@example.com>` which wasn't fun for users to type and involved a lot of splitting/unsplitting strings.
3. Nonstandard label in the option_value table contained `<` and `>` characters, making general rules about html escaping difficult.

After
--------
New table has just the fields needed for storing site email addresses. UI switched to SearchKit/FormBuilder:

![image](https://github.com/user-attachments/assets/eacaa832-4191-4636-b22a-37228fd6e74e)

colemanw committed:
Add new table for SiteEmailAddresses

Introduce a new entity type `SiteEmailAddress` as a replacement
to the 'from_email_address' option group.

totten commented:
Personal feeling: __yes, please, for the love of god, yes.__ 

---

Big consideration is going to be technical interop for extensions and sysadmin scripts. (*In my copy of `universe`, there are ~90 references to `from_email_address` in ~45 extensions. That's a mix of things, and some may not be OG-references. But a lot are*) Maybe there's some clever mechanism to help smooth it over? 

(Like... `hook_optionValues`?  Or, anticipating a mix of reads+writes and BAO/APIv3/APIv4, it may merit a bidirectional sync -- eg an extension `legacyemailsync` to uses `hook_post` to sync the tables. I don't really know.)

---

In the future, we may get some better options to address permissions and routing if it's a proper entity.

---

Small aside `#bikeshed`: in `ezcMailAddress` and `Civi\WorkflowMessage\Traits\AddressingTrait`, they refer to these properties as `name`,`email` not `display_name`,`email`. I think `name` is probably more common in email-land, so I would gently vote for that. Not a huge deal.

highfalutin commented:
I don't deal with _from_email_address_ much, but the proposed schema looks good to me. What would be the best practice in Civi for maintaining a single `is_default = 1` per domain?

colemanw commented:
> @highfalutin What would be the best practice in Civi for maintaining a single is_default = 1 per domain?

I added a unique index to the table for that, and the business logic would include an implementation of `hook_civicrm_pre` to set all others in the domain to `false` whenever setting one to `true`.

colemanw commented:
> @totten  I think name is probably more common in email-land, so I would gently vote for that. Not a huge deal.

Well in Civi-land, `name` usually means "machine name" so I avoided it for that reason, but since there's precedent with other email entities, sure let's go with `name`.

totten commented:
> Well in Civi-land, name usually means "machine name" so I avoided it for that reason, but since there's precedent with other email entities, sure let's go with name.

Ah, good point. Terms are painted into a corner. Dealers choice.

colemanw committed:
Refactor from_email_address to SiteEmailAddress API.

Replaced legacy "from_email_address" option group references with the new API.

colemanw committed:
SiteEmailAddress - Enforce default one default per domain

Also removed unnecessary domain filtering logic from OptionValue as this was the only option group it applied to.

colemanw committed:
SiteEmailAddress - Add legacy API adapter

Wraps the OptionValue api (v3 and v4) to provide backward compatibility
for accessing SiteEmailAddress via the OptionValue api.
CRM_Core_OptionGroup::values and ::getDefaultValue are also supported.

colemanw committed:
SiteEmailAddress - Add search display and formbuilder form

Also removes unused handling from the optionValue quickform

mattwire commented on CRM/Upgrade/Incremental/schema/6.0.alpha1.DomainEmailAddress.entityType.php:
```diff
@@ -0,0 +1,108 @@
+      'sql_type' => 'varchar(63)',
+      'input_type' => 'Text',
+      'required' => TRUE,
+      'description' => ts('Full name of the recipient.'),
```
```suggestion
      'description' => ts('Full name of the Fromage.'),
```
Well it's not the recipient as it's the "Text that is displayed instead of the email address from which the email comes from" but I couldn't work out how to make that sound sensible so I just changed it to the French for Cheese instead.

colemanw replied:
Very helpful suggestion, although I find the wording a bit cheesy.

mattwire commented on schema/Core/DomainEmailAddress.entityType.php:
```diff
@@ -0,0 +1,108 @@
+      'sql_type' => 'varchar(63)',
+      'input_type' => 'Text',
+      'required' => TRUE,
+      'description' => ts('Full name of the recipient.'),
```
Umm, see my earlier comment about this description. Before I write something else about cheese.

Sjord commented on CRM/Core/BAO/Domain.php:
```diff
@@ -192,19 +192,18 @@ public static function multipleDomains() {
       if ($returnString) {
         // Return a string like: "Demonstrators Anonymous" <info@example.org>
-        return $fromEmailAddress;
+        return [CRM_Utils_Mail::formatFromAddress($fromEmailAddress)];
```
This returns an array with a single element. Is that intentional? The comment and variable name imply that a string is returned, not an array.

colemanw replied:
The 2nd param to this function was pretty weird & the comments weren't very helpful.
I've pushed a commit to deprecate it.

Sjord commented on CRM/Core/OptionGroup.php:
```diff
@@ -611,4 +599,30 @@ public static function flushAll() {
+    }
+    if ($condition && is_string($condition)) {
+      // Use regex to extract the field_name and the value from the condition.
+      if (preg_match("/([\w_]+)\s*=\s*(\d+)$/", $condition, $matches)) {
```
Currently, the condition is ignored if it does not match the regex. Is that ok, or should it throw an error if the condition is an unexpected format?

colemanw replied:
Thanks for the suggestion. I agree & have updated the PR.

Sjord commented on CRM/Utils/Mail.php:
```diff
@@ -649,12 +649,16 @@ public static function format($fields) {
    */
   public static function formatFromAddress($from) {
+    if (is_array($from)) {
+      return "\"{$from['name']}\" <{$from['email']}>";
```
Should quotes or angular brackets be encoded or escaped here?

colemanw replied:
This function returns an unescaped value. See a few lines down.

Sjord commented on CRM/Core/BAO/Email.php:
```diff
@@ -288,9 +288,14 @@ public static function holdEmail(array &$email) {
+      ->execute();
+    foreach ($domainFrom as $address) {
+      $domainEmail = CRM_Utils_Mail::formatFromAddress($address);
       $domainEmails[$domainEmail] = htmlspecialchars($domainEmail);
```
I wouldn't expect `htmlspecialchars` here. I would expect this to return the emails as-is, and then it's the template's job to escape them. It seems this was already here, so maybe this should remain in place for compatibility reasons.

colemanw committed:
SiteEmailAddress - Deprecate 2nd param to CRM_Core_BAO_Domain::getNameAndEmail

Deprecated the `returnFormatted` parameter in `getNameAndEmail` and updated relevant code to align with the new method.

eileenmcnaughton commented:
Generally good - I'm actually -1 on using `name` for anything other than machine name - and I feel like I might have a use case for adding `name` as in machine name to the table - but haven't fully thought it through - we certainly have different from addresses for different purposes & use different from addresses with different templates - which is what machine name would often be used to handle. If we don't add `name` as in machine name then I have a feeling there is a requirement for some extra field that will come up fairly quickly once this is done because currently the option value structure is so nasty no-one asks for it to be improved on....

eileenmcnaughton commented:
ALSO - I believe this option group has prevented us from passing optionValue through html_entities....

colemanw commented:
Ok @eileenmcnaughton has convinced me to go back to `display_name`. Sorry @totten you're outvoted.

totten commented on schema/Core/DomainEmailAddress.entityType.php:
```diff
@@ -0,0 +1,116 @@
+  'getInfo' => fn() => [
+    'title' => ts('Site Email Address'),
+    'title_plural' => ts('Site Email Addresses'),
+    'description' => ts('Return addresses to use for system-generated emails.'),
```
That's misleading, isn't it?

* The options here are literally the `From:` addresses that the staff-user can choose when composing "New Email (*activity*)" or "New Mailing". That's not system-generated.
* Maybe... if an address is flagged `is_default=1`, then that could be used for system-generated emails?
    * (*But I'm really not sure about that -- eg for better or worse, for `Event`s and `ContributionPage`s, the system-generated confirmations are actually based on different `From`.*)
* Emails may or may-not have a return-channel. (And there's a common case -- CiviMail -- where the return-channel differs distinctly from the `From:` address.)
* The decision to register addresses here is pretty subjective... depending on... attitudes about branding, mail-server configuration, usage of `allow_mail_contact_email`, etc.

I don't care too much about perfecting help-text that only appears to devs. But... that description is very different from my understanding of the data. Are there actually multiple scenarios flowing this data? And therefore miscommunication about its purpose?

colemanw replied:
@totten good catch. I've changed the wording to 'Sender addresses to use for outbound emails.' which should be less confusing.

totten commented on schema/Core/DomainEmailAddress.entityType.php:
```diff
@@ -0,0 +1,116 @@
+    'browse' => 'civicrm/admin/options/from_email_address',
+    'add' => 'civicrm/admin/form/site-email-address',
+    'update' => 'civicrm/admin/form/site-email-address#?email=[id]',
+  ],
```
Hmm, sensible functionality, but this highlights some confusion in naming:

* __Before__: one phrase, "_from email address_"
* __After__: three phrases, varying by context:
    * Sometimes "_domain email address_" (eg APIv4 and SQL)
    * Sometimes "_site email address_" (eg new routes)
    * Sometimes "_from email address_" (eg existing routes)

Each name has its own semantic niggles, so I can share some ambivalence. And TBH, I hardly noticed that the initial draft  spoke to `from_email_address` and `DomainEmailAddress`. (The option-group namespace differs from the APIv4 namespace, so sure... there's gonna be something different there.)

But three different names? That's gonna be confusing later, right?

(If there is some ambiguity for the name of this `$thing`, then I guess... the protocol is to talk that through and pick one?)

ufundo replied:
+1 for switching to Site Email Address everywhere

colemanw replied:
Agree with @ufundo. I consider "Domain" to be techie speak for "Website" or just "Site" so this PR as-is moves us in the direction of:

- **Entity/Api name:** DomainEmailAddress
- **User-facing name:** Site Email Address

All that's left is to cleanup the old references to "from_email_address"

totten commented:
On the auto-build demo site, I tried registering a new address. The admin GUI seemed to work properly. :tada: 

But then I went to "Mailings => New Mailing", and the page crashes:

![Screenshot from 2025-02-05 21-35-34](https://github.com/user-attachments/assets/99cb75f2-7a76-4d4b-84a1-3ca3def8eaa1)

totten commented:
> (@eileenmcnaughton)  I feel like I might have a use case for adding `name` as in machine name to the table

Yeah, fair enough. We can do `display_name` here. But just looking toward the `name` contingency...  the existing `OptionGroup`/`OptionValue` already has machine-name. And with the way it works, it tends to become [misleading gibberish](https://github.com/user-attachments/assets/6ca8322b-c8f4-4281-82bf-9f816b3620dd) -- i.e. your typical/default record winds up with `name="FIXME" <info@EXAMPLE.ORG>` (which is uneditable and which contradicts the real data in the record).

There may exist some good use-case for machine `name` on an email address. I don't know, but assuming it exists... I'd probably set the default/computed name to something like `md5(random_bytes(16))` (*which may also be gibberish, but at least it's truthful gibberish :upside_down_face:*). In that arrangement, someone using dev-tools could still assign a mnemonic machine-name for records they care about.

totten commented:
OK, so I did some light/local r-run with an upgrade scenario. The idea is:

* Install 5.81
    * Add several addresses and exercise the various options (default, disabled, description, etc)
    * Click around some UI's
    * Call some of those common APIs. Dump output to a file.
* Upgrade to PR
    * Click around some UI's
    * Call some of those common APIs. Dump output to a file. Compare.

The main thing that stuck to me is the change re: ordering/weights. This isn't mentioned in the "Description", but (in usage) it was apparent on both the UI level and API level.

* UI: Before
    > ![Screenshot from 2025-02-06 03-07-15](https://github.com/user-attachments/assets/6ddf6d9a-8a9b-418e-bc7c-8caa86b8bd1d)
* UI: After
    > ![Screenshot from 2025-02-06 03-07-27](https://github.com/user-attachments/assets/8c1ac068-a2b9-4dcc-927e-fb7acb617d04)

* To me, the old UI feels overwrought -- I would feel silly fiddling with the `weight`s on a record-by-record basis. Alphabetical ordering would be much better balance between (a) providing predictable behavior and (b) not-wasting-time-with-clickery-clackery.

    So to my mind: ___woot! let's remove `weight`! it's an anti-feature!___

* OTOH, order/weight has been there forever; it's a common idiom in Civi-land; the API-mapping fails if the API-call involves `weight`; and there's usually some kind of history for why a feature like that exists. My main lament is that we have no good way to identify/learn-from the user who actually likes the `weight` functionality. But I guess if we can get a couple dev's nodding in agreement that `weight` is a terrible idea, then that'll be enough...

* Spot-checking how the ordering actually behaves in a few UIs:

    * __"View Contact > Contribution > Send Receipt > Email"__: Works beautifully. Alphabetical. Default respected. Disable/enable flag respected.
    * __"New Email (activity)"__: The items are alphabetical. However, it doesn't respect the `is_default` flag. 
        * This may not be obvious if your configuration gives precedence to personal emails (`allow_mail_from_logged_in_contact=1`).  However, the suitability of that flag is *very* subjective (*depending on both marketing preferences and mail-server config*). In the safe case of `allow_mail_from_logged_in_contact=0`, the default email addr is particularly important.
    * __"New Mailing"__: Crashes, so can't say.
    * __"Administer > Communications > From Email Addresses"__: Doesn't sort by default

ufundo commented:
**+1 for Site Email Address over Domain Email Address or From Email Address** (commented in a thread above but just collating here) 

**+1 for display_name**

**Machine `name` field**

I'm with @eileenmcnaughton that there could be all kinds of uses for machine named records.

> @totten  it tends to become [misleading gibberish](https://github.com/user-attachments/assets/6ca8322b-c8f4-4281-82bf-9f816b3620dd) -- i.e. your typical/default record winds up with name="FIXME" <info@EXAMPLE.ORG>

I don't dislike that auto-generated machine `name`s correspond to the initial `display_name` and `email`, and may deviate from it if the display name is updated. That's a standard behaviour in lots of places where we have `name` - `label`. Yes if you create a record with label = `Red`, name = `red` it's unfortunate that if you change the label to `Blue` your name is still `red`. But it's often useful evidence trail if you're coming in to try and diagnose what an overambitious admin has done to their config. And it's a human-level CRM problem for when it's appropriate to edit an existing record versus delete it and add a new one. The name behaviour is useful to keep name red if you change the label from `Red` to `Rouge`.  In your "misleading gibberish" screenshot, I think its a feature that you can seee the record now known as Foo Foo is the originally (and persistently) named as Foo. There's a bit of subtlety here - but that's not inappropriate for a below-the-hood feature.

I do think it's unfortunate that you get FIXME in the starting one. Could we hard code a value for that one "initial"?

I also think it would be good (and consistent) to munge what goes into the `name` field (ie at least replace special chars with `_`, and maybe use some conventional separator for the display_name and the email?

Having said all this - it looks like adding a `name` field could be a separate follow up PR? @colemanw was that you thinking

colemanw commented:
> @ufundo it looks like adding a name field could be a separate follow up PR?

Definitely! I'm not about to make this PR any bigger than it already is.

> @totten So to my mind: woot! let's remove weight! it's an anti-feature!

Exactly. Also, just to make the obvious point, `weight` was never intentionally added to this option group, it was _inherited_ from the OptionValue table, just as the overwrought UI in your screenshot was inherited from the generic "edit option values" screen. This new table is more intentional about only including columns that are actually relevant.

Re the `is_default` flag being respected. I'll look but that might be a preexisting bug. None of the legacy functions I adapted were doing any sorting by `is_default`.

colemanw commented:
I've rebased/pushed this PR to address the following

1. **CiviMail page crash:** This was due to incomplete refactoring of the `crmFromAddresses` angular service. I've decided to revert my refactoring and instead put more work into the api compat layer to ensure smooth backward-compat with things like `weight`. Tests added. Mailing screen now works as-before, and switching it to use the new api can wait for another day (it's easier said than done because that core angular code is reused by some extensions in universe).
2. **Entity description** reworded to 'Sender addresses to use for outbound emails.' per @totten's feedback.
3. **Administer > Communications > From Email Addresses:** Default sort added to show default first, then sort alphabetically.

I also looked into @totten's testing of the is_default flag behavior on email-related screens. I agree it's not working as expected in this PR, but then I build a vanilla site from master and got the exact same behavior. So the not-workingness is a pre-existing condition and I'm happy to take a look at it but am wary of bogging down this PR even more so I [opened a new issue for that bug](https://lab.civicrm.org/dev/core/-/issues/5720).

colemanw committed:
SiteEmailAddress - Standardize language

Consistently use "Site Email Addresses" instead of "FROM Email Addresses"
(and no need to SHOUT).

totten committed:
Leave XML stub for old civicrm/admin/options/from_email_address

totten committed:
Regenerate sql/civicrm_generated.mysql

totten committed:
Upgrader - Snapshot option-values before deleting them

totten commented:
Re-tested. Added a couple more tweaks. Merge on pass.
